dace/sdfg/propagation.py

```diff
@@ -674,7 +673,194 @@ def _annotate_loop_ranges(sdfg, unannotated_cycle_states):
+    reset_state_annotations(sdfg)
+
+    # Clean up the state machine by separating combined condition and assignment
+    # edges.
+    split_interstate_edges(sdfg)
+
+    # To enable branch annotation, we add a temporary exit state that connects
+    # to all child-less states. With this, we can use the dominance frontier
+    # to determine a full-merge state for branches.
+    temp_exit_state = None
+    for s in sdfg.nodes():
+        if sdfg.out_degree(s) == 0:
+            if temp_exit_state is None:
+                temp_exit_state = sdfg.add_state('__dace_brannotate_exit')
+            sdfg.add_edge(s, temp_exit_state, InterstateEdge())
+
+    dom_frontier = cfg.acyclic_dominance_frontier(sdfg)
+
+    # Find any valid for loop constructs and annotate the loop ranges. Any other
+    # cycle should be marked as unannotated.
+    unannotated_cycle_states = []
+    _annotate_loop_ranges(sdfg, unannotated_cycle_states)
+
+    # Keep track of states that fully merge a previous conditional split. We do
+    # this so we can remove the dynamic executions flag for those states.
+    full_merge_states = set()
+
+    visited_states = set()
+
+    traversal_q = deque()
+    traversal_q.append((sdfg.start_state, 1, False, []))
+    while traversal_q:
+        (state, proposed_executions, proposed_dynamic, itvar_stack) = traversal_q.pop()
+
+        out_degree = sdfg.out_degree(state)
+        out_edges = sdfg.out_edges(state)
+
+        # Check if the traversal reached a state that's already been visited
+        # (ends traversal), or if the number of executions being propagated is
+        # dynamic unbounded. Otherwise, continue regular traversal.
+        if state in visited_states:
+            # This state has already been visited.
+            if getattr(state, 'is_loop_guard', False):
+                # If we encounter a loop guard that's already been visited,
+                # we've finished traversing a loop and can remove that loop's
+                # iteration variable from the stack. We additively merge the
+                # number of executions.
+                state.executions += proposed_executions
+            else:
+                # If we have already visited this state, but it is NOT a loop
+                # guard, this means that we can reach this state via multiple
+                # different paths. If so, the number of executions for this
+                # state is given by the maximum number of executions among each
+                # of the paths reaching it. If the state additionally completely
+                # merges a previously branched out state tree, we know that the
+                # number of executions isn't dynamic anymore.
+                state.executions = sympy.Max(state.executions, proposed_executions).doit()
+                if state in full_merge_states:
+                    state.dynamic_executions = False
+                # TODO: do we need this else here or not?
+                # else:
```

Comment:
Yes there are some situations where this may be needed. Consider an if-else state machine which merges back into some common state `mergestate`, but where one of the branches may terminate the program based on some additional condition. In this case `mergestate` is not a full merge state, since there are program paths that may not reach it. In that case, the number of executions is dynamic, i.e. in `[0:state.executions]`.